crates/trie-parallel/src/parallel_root.rs

```diff
@@ -0,0 +1,262 @@
+    trie::{HashBuilder, Nibbles, TrieAccount},
+    B256,
+};
+use reth_provider::{
+    providers::{ConsistentDbView, ConsistentViewError},
+    DatabaseProviderFactory, ProviderError,
+};
+use reth_trie::{
+    hashed_cursor::{HashedCursorFactory, HashedPostStateCursorFactory},
+    node_iter::{AccountNode, AccountNodeIter},
+    trie_cursor::TrieCursorFactory,
+    updates::TrieUpdates,
+    walker::TrieWalker,
+    HashedPostState, StorageRoot, StorageRootError,
+};
+use std::collections::HashMap;
+use thiserror::Error;
+use tracing::*;
+
+/// Parallel incremental state root calculator.
+///
+/// The calculator starts off by pre-computing storage roots of changed
+/// accounts in parallel. Once that's done, it proceeds to walking the state
+/// trie retrieving the pre-computed storage roots when needed.
+///
+/// Internally, the calculator uses [ConsistentDbView] since
+/// it needs to rely on database state saying the same until
+/// the last transaction is open.
+/// See docs of using [ConsistentDbView] for caveats.
+///
+/// If possible, use more optimized `AsyncStateRoot` instead.
+#[derive(Debug)]
+pub struct ParallelStateRoot<DB, Provider> {
+    /// Consistent view of the database.
+    view: ConsistentDbView<DB, Provider>,
+    /// Changed hashed state.
+    hashed_state: HashedPostState,
+}
+
+impl<DB, Provider> ParallelStateRoot<DB, Provider> {
+    /// Create new parallel state root calculator.
+    pub fn new(view: ConsistentDbView<DB, Provider>, hashed_state: HashedPostState) -> Self {
+        Self { view, hashed_state }
+    }
+}
+
+impl<DB, Provider> ParallelStateRoot<DB, Provider>
+where
+    DB: Database,
+    Provider: DatabaseProviderFactory<DB> + Send + Sync,
+{
+    /// Calculate incremental state root in parallel.
+    pub fn incremental_root(self) -> Result<B256, ParallelStateRootError> {
+        self.calculate(false).map(|(root, _)| root)
+    }
+
+    /// Calculate incremental state root with updates in parallel.
+    pub fn incremental_root_with_updates(
+        self,
+    ) -> Result<(B256, TrieUpdates), ParallelStateRootError> {
+        self.calculate(true)
+    }
+
+    fn calculate(
+        self,
+        retain_updates: bool,
+    ) -> Result<(B256, TrieUpdates), ParallelStateRootError> {
+        let prefix_sets = self.hashed_state.construct_prefix_sets();
+        let storage_root_targets = StorageRootTargets::new(
+            self.hashed_state.accounts.keys().copied(),
+            prefix_sets.storage_prefix_sets,
+        );
+        let hashed_state_sorted = self.hashed_state.into_sorted();
+
+        // Pre-calculate storage roots in parallel for accounts which were changed.
+        debug!(target: "trie::parallel_state_root", len = storage_root_targets.len(), "pre-calculating storage roots");
+        let mut storage_roots = storage_root_targets
+            .into_par_iter()
+            .map(|(hashed_address, prefix_set)| {
+                let provider_ro = self.view.provider_ro()?;
```

Comment:
is this even necessary?
since Tx: Send+ Sync you should be able to access this in this scope so you only need to create it once

although the tx ops will sync via mutex

Reply:
absolutely, because the calculations will choke on that mutex